File: func-pipes/static/js/power2/nodes/series-battery.js
/*
  nodes/series-battery.js — Rechargeable Battery / Supercapacitor
  ─────────────────────────────────────────────────────────────────────────────
  A rechargeable battery with both an inbound (charging) pip and an outbound
  (output) pip. It operates independently of the inbound voltage — it always
  emits its own rated EMF (panel.volts) downstream. The inbound signal purely
  charges the energy store; the outbound load draws from it.

  Energy model
  ────────────
  Each tick:
    chargeInW   = V_in × min(A_in, chargeAmps)   — power arriving from charger
    chargeOutW  = drawWatts                        — actual downstream load (BFS)
    net (W)     = chargeInW − chargeOutW
    ΔchargeWh   = net × dt / 3600

  If chargeWh > 0 the battery emits its own { v: panel.volts, a: panel.amps }
  regardless of whether anything is connected to the inbound pip.
  When chargeWh reaches 0 the battery goes 'dead' and emits null.
  It auto-revives as soon as enough charge has accumulated (> 1% capacity).

  States: 'off' | 'charging' | 'discharging' | 'full' | 'dead'

  Extended state
  ──────────────
  live          bool    — output enabled; false = no EMF out
  volts         number  — rated output voltage (default 12)
  amps          number  — rated max output current (default 20)
  chargeAmps    number  — max inbound charge current accepted (default 10)
  capacityWh    number  — total energy capacity in watt-hours (default 10)
  chargeWh      number  — current stored energy
  chargePercent number  — 0–100 live readout
  drawWatts     number  — downstream load draw, updated by BFS each frame
  chargeInW     number  — live inbound charge power (W)
  chargeOutW    number  — live outbound draw power (W)
  inVolts       number  — live inbound voltage readout
  inAmps        number  — live inbound current readout
*/

class SeriesBattery extends NodeBase {

    static type  = 'series-bat'
    static label = 'Battery'
    static group = 'Source'

    static catalog = [
        { key: 'series-12v',   label: 'Battery 12V 10Ah',   volts: 12,  amps: 20,  chargeAmps: 10, capacityWh: 120  },
        { key: 'series-24v',   label: 'Battery 24V 10Ah',   volts: 24,  amps: 20,  chargeAmps: 10, capacityWh: 240  },
        { key: 'series-48v',   label: 'Battery 48V 20Ah',   volts: 48,  amps: 30,  chargeAmps: 15, capacityWh: 960  },
        { key: 'series-lipo',  label: 'LiPo 3.7V  5Ah',     volts: 3.7, amps: 10,  chargeAmps: 5,  capacityWh: 18.5 },
        { key: 'series-9v',    label: 'PP3  9V  0.5Ah',      volts: 9,   amps: 1,   chargeAmps: 0.5,capacityWh: 4.5  },
        { key: 'series-super', label: 'Supercap 12V 0.1Wh',  volts: 12,  amps: 50,  chargeAmps: 50, capacityWh: 0.1  },
    ]

    static _defaultRipple() {
        return { enabled: false, amount: 0.5, interval: 0.5 }
    }

    static _defaultSpike() {
        return { enabled: true, percent: 10, duration: 0.3 }
    }

    static defaults(id, preset = {}) {
        const cap = preset.capacityWh ?? 10
        return {
            ...super.defaults(id, preset),
            label:         preset.label      || 'Battery',
            volts:         preset.volts      ?? 12,
            amps:          preset.amps       ?? 20,
            chargeAmps:    preset.chargeAmps ?? 10,
            capacityWh:    cap,
            chargeWh:      cap,
            chargePercent: 100,
            drawWatts:     0,
            chargeInW:     0,
            chargeOutW:    0,
            inVolts:       0,
            inAmps:        0,
            live:          true,
            ripple:        preset.ripple ? { ...preset.ripple } : { ...this._defaultRipple() },
            spike:         preset.spike  ? { ...preset.spike  } : { ...this._defaultSpike()  },
        }
    }

    static configFields() {
        return [...super.configFields(), 'volts', 'amps', 'chargeAmps', 'capacityWh', 'live', 'ripple', 'spike']
    }

    /**
     * Process an inbound (charging) signal. When live, adds the battery's own
     * rated EMF on top of the inbound voltage and forwards downstream. When
     * not live, forwards the inbound signal unchanged (pass-through mode).
     * Emits null if the battery is dead or has no charge.
     * @param {Object}     panel
     * @param {Object|null} signal — upstream { v, a } or null
     * @param {PowerGraph} graph
     */
    static apply(panel, signal, graph) {
        // Store inbound signal for charge calculation in tick
        panel.inVolts = signal ? signal.v : 0
        panel.inAmps  = signal ? signal.a : 0

        if (panel.state === 'dead') {
            graph.emit(panel, null)
            return
        }

        const prev = panel.state

        if (!panel.live) {
            // Pass-through: forward inbound signal unchanged, no charge used
            if (signal && signal.v > 0) {
                panel.state = 'pass'
                graph.emit(panel, { v: signal.v, a: Math.min(signal.a, panel.amps) })
            } else {
                panel.state = 'off'
                graph.emit(panel, null)
            }
            if (panel.state !== prev)
                SeriesBattery.dispatch(panel, 'state:change', { from: prev, to: panel.state })
            return
        }

        if (panel.chargeWh <= 0) {
            panel.state    = 'dead'
            panel.chargeWh = 0
            SeriesBattery.dispatch(panel, 'battery:dead', { chargePercent: 0 })
            SeriesBattery.dispatch(panel, 'state:change', { from: prev, to: 'dead' })
            graph.emit(panel, null)
            graph.updateAllGenDraws()
            return
        }

        // Emit own rated EMF stacked on top of inbound voltage
        const vIn  = signal ? signal.v : 0
        const aOut = signal ? Math.min(signal.a, panel.amps) : panel.amps
        const vOff = panel._rippleOffset ?? 0
        const m    = NodeBase.spikeMultiplier(panel)
        graph.emit(panel, { v: (vIn + panel.volts + vOff) * m, a: aOut * m })
    }

    /**
     * Per-frame energy accounting. Decays any inrush spike, then independently
     * computes charge-in (from inbound source) and charge-out (proportional
     * share of downstream load) and updates chargeWh. Transitions between
     * charging / discharging / full / dead states and revives the battery once
     * enough charge has accumulated after a dead event.
     * @param {Object}     panel
     * @param {number}     dt    — elapsed seconds since last tick
     * @param {PowerGraph} graph
     */
    static tick(panel, dt, graph) {
        // Decay inrush spike and re-apply to settle downstream on expiry too.
        const wasNonZero = (panel._spikeTimer ?? 0) > 0
        const active     = NodeBase.tickSpike(panel, dt)
        if ((active || (wasNonZero && !active)) &&
                panel.live && panel.state !== 'dead' && panel.signal !== undefined)
            SeriesBattery.apply(panel, panel.signal, graph)

        // Charging and discharging are independent processes.
        //
        // Charge in: inbound source pushes current into the battery at up to
        //   chargeAmps, stored at the battery's own voltage.
        //   chargeInW = min(A_in, chargeAmps) × V_bat
        //
        // Charge out: the battery's proportional share of the downstream load.
        //   In a series stack, each source's share ∝ its voltage fraction.
        //   chargeOutW = drawWatts × V_bat / (V_in + V_bat)
        //   When standalone (V_in = 0) this correctly equals drawWatts.

        const totalV     = panel.inVolts + panel.volts
        const chargeInW  = (panel.inVolts > 0)
            ? Math.min(panel.inAmps, panel.chargeAmps) * panel.volts
            : 0
        const chargeOutW = (panel.live && totalV > 0 && panel.drawWatts > 0)
            ? panel.drawWatts * (panel.volts / totalV)
            : 0

        panel.chargeInW  = +chargeInW.toFixed(1)
        panel.chargeOutW = +chargeOutW.toFixed(1)

        const netW = chargeInW - chargeOutW
        panel.chargeWh = Math.min(
            panel.capacityWh,
            Math.max(0, panel.chargeWh + (netW * dt) / 3600)
        )
        panel.chargePercent = +(panel.chargeWh / panel.capacityWh * 100).toFixed(1)

        const prevState = panel.state

        if (panel.state === 'dead') {
            if (panel.chargeWh >= panel.capacityWh * 0.01 && chargeInW > 0) {
                panel.state = 'charging'
                panel.live  = true
                SeriesBattery.dispatch(panel, 'battery:revived', { chargePercent: panel.chargePercent })
                SeriesBattery.dispatch(panel, 'state:change', { from: 'dead', to: 'charging' })
                SeriesBattery.apply(panel, panel.signal, graph)
                graph.updateAllGenDraws()
            }
            return
        }

        if (panel.chargeWh <= 0) {
            panel.state = 'dead'
            panel.live  = false
            SeriesBattery.dispatch(panel, 'battery:dead', { chargePercent: 0 })
            SeriesBattery.dispatch(panel, 'state:change', { from: prevState, to: 'dead' })
            graph.emit(panel, null)
            graph.updateAllGenDraws()
            return
        }

        if (!panel.live) { panel.state = panel.signal?.v > 0 ? 'pass' : 'off'; return }

        if (panel.chargeWh >= panel.capacityWh) {
            panel.state = 'full'
        } else if (netW >= 0) {
            panel.state = 'charging'
        } else {
            panel.state = 'discharging'
        }

        if (panel.state !== prevState)
            SeriesBattery.dispatch(panel, 'state:change', { from: prevState, to: panel.state })

        // Charge telemetry — throttled, dispatched only when value changes
        const pct = panel.chargePercent
        if (pct !== panel._lastChargePct) {
            panel._lastChargePct = pct
            SeriesBattery.throttle(panel, 'battery:charge', {
                chargePercent: pct,
                chargeWh:      +panel.chargeWh.toFixed(3),
                chargeInW:     panel.chargeInW,
                chargeOutW:    panel.chargeOutW,
            })
        }
    }

    // ── Actions ───────────────────────────────────────────────────────────────

    /**
     * Toggle battery output on or off. No-op if the battery is dead —
     * call reset() first to restore it. Starts an inrush spike when enabling.
     * @param {Object}     panel
     * @param {PowerGraph} graph
     */
    static toggle(panel, graph) {
        if (panel.state === 'dead') return   // must reset first
        panel.live = !panel.live
        if (panel.live) NodeBase.startSpike(panel)
        SeriesBattery.dispatch(panel, 'battery:toggle', { live: panel.live })
        SeriesBattery.apply(panel, panel.signal, graph)
        graph.updateAllGenDraws()
    }

    /**
     * Toggle between boost mode (live=true, adds own EMF) and pass-through
     * mode (live=false, forwards inbound signal with no voltage addition).
     * No-op if the battery is dead.
     * @param {Object}     panel
     * @param {PowerGraph} graph
     */
    static togglePass(panel, graph) {
        if (panel.state === 'dead') return
        if (!panel.live) {
            // already in pass-through → switch back to boost
            panel.live = true
        } else {
            panel.live = false
        }
        SeriesBattery.dispatch(panel, 'battery:pass-toggle', { live: panel.live })
        SeriesBattery.apply(panel, panel.signal, graph)
        graph.updateAllGenDraws()
    }

    /**
     * Validate and re-apply after external config changes (e.g. capacityWh or
     * chargeAmps edited via the config panel). Clamps chargeWh to capacity.
     * @param {Object}     panel
     * @param {PowerGraph} graph
     */
    static paramsChanged(panel, graph) {
        if (panel.chargeWh > panel.capacityWh) panel.chargeWh = panel.capacityWh
        panel.chargePercent = +(panel.chargeWh / panel.capacityWh * 100).toFixed(1)
        SeriesBattery.apply(panel, panel.signal, graph)
        graph.updateAllGenDraws()
    }

    /**
     * Full reset — restores the battery to 100% charge, re-enables output,
     * clears all telemetry accumulators, and delegates to NodeBase.reset().
     * @param {Object}     panel
     * @param {PowerGraph} graph
     */
    static reset(panel, graph) {
        panel.live          = true
        panel.chargeWh      = panel.capacityWh
        panel.chargePercent = 100
        panel.drawWatts     = 0
        panel.chargeInW     = 0
        panel.chargeOutW    = 0
        panel.inVolts       = 0
        panel.inAmps        = 0
        panel._lastChargePct = null
        SeriesBattery.dispatch(panel, 'battery:reset', { chargePercent: 100 })
        super.reset(panel, graph)
    }
}

NodeRegistry.register(SeriesBattery)

File: power_graph/src/power_graph/nodes/series_battery.py
"""
nodes/series_battery.py
──────────────────────────────────────────────────────────────────────────────
SeriesBattery — rechargeable battery in a series circuit.

States: 'off' | 'charging' | 'discharging' | 'full' | 'dead' | 'pass'

  pass        — passthrough only, no stored charge (chargeAmps = 0 or cap = 0)
  charging    — receiving power, storing energy
  discharging — providing own EMF contribution when downstream draws current
  full        — at capacity; stops charging
  dead        — charge exhausted; no contribution
  off         — no inbound signal and dead/empty

Energy model
  chargeInW  = min(inAmps, chargeAmps) * volts          (watts into battery)
  chargeOutW = drawWatts * (volts / totalV)              (battery share)
  chargeWh   += chargeInW  * dt / 3600                  (per tick)
  chargeWh   -= chargeOutW * dt / 3600                  (per tick, when discharging)

The battery stacks its own EMF onto the inbound signal voltage so series
batteries increase total line voltage.
"""

from typing import Dict, List
from power_graph.node_base import NodeBase, Signal
from power_graph.node_registry import NodeRegistry


class SeriesBattery(NodeBase):
    """Rechargeable series-circuit battery."""

    type  = 'series-battery'
    label = 'Battery'
    group = 'Storage'
    dispatch_delay = 200

    @classmethod
    def _default_ripple(cls):
        # Internal resistance variation under charge/discharge cycles
        return RippleProfile(enabled=True, amount=0.6, interval=1.2)

    catalog = [
        {'key': 'series-12v',    'label': 'Lead-Acid 12V 7Ah',  'volts': 12,  'amps': 7,   'chargeAmps': 2,  'capacityWh': 84 },
        {'key': 'series-24v',    'label': 'Lead-Acid 24V 7Ah',  'volts': 24,  'amps': 7,   'chargeAmps': 3,  'capacityWh': 168},
        {'key': 'series-48v',    'label': 'LiFePO4 48V 10Ah',   'volts': 48,  'amps': 10,  'chargeAmps': 5,  'capacityWh': 480},
        {'key': 'series-lipo',   'label': 'LiPo 11.1V 5Ah',     'volts': 11.1,'amps': 5,   'chargeAmps': 3,  'capacityWh': 55.5},
        {'key': 'series-9v',     'label': '9V Block',            'volts': 9,   'amps': 0.6, 'chargeAmps': 0,  'capacityWh': 5.4},
        {'key': 'series-super',  'label': 'Supercapacitor 2.7V', 'volts': 2.7, 'amps': 50,  'chargeAmps': 50, 'capacityWh': 0.05},
    ]

    @classmethod
    def defaults(cls, node_id: int, preset: Dict = None) -> Dict:
        if preset is None:
            preset = {}
        base = super().defaults(node_id, preset)
        cap_wh = preset.get('capacityWh', 84)
        base.update({
            'volts':          preset.get('volts',      12),
            'amps':           preset.get('amps',       7),
            'chargeAmps':     preset.get('chargeAmps', 2),
            'capacityWh':     cap_wh,
            'chargeWh':       preset.get('chargeWh', cap_wh * 0.8),  # start at 80%
            'chargePercent':  80.0,
            'drawWatts':      0.0,
            'chargeInW':      0.0,
            'chargeOutW':     0.0,
            'inVolts':        0.0,
            'inAmps':         0.0,
            'live':           False,
            'running':        True,   # internal on/off switch (toggled by user)
            'state':          'off',
        })
        return base

    @classmethod
    def config_fields(cls) -> List[str]:
        return [*super().config_fields(), 'volts', 'amps', 'chargeAmps', 'capacityWh']

    @classmethod
    def toggle(cls, panel: Dict, graph):
        """Internal on/off switch — independent of graph enabled state."""
        panel['running'] = not panel.get('running', True)
        prev = panel['state']
        if not panel['running']:
            panel['live']  = False
            panel['state'] = 'off'
            cls.dispatch(panel, 'state:change', {'from': prev, 'to': 'off'})
            graph.emit(panel, None)
        else:
            # Re-apply with current sources so battery resumes immediately
            sources  = panel.get('powerSources', {})
            combined = graph.combine_sources(sources) if sources else None
            cls.apply(panel, combined, graph)

    @classmethod
    def apply(cls, panel: Dict, signal: Signal, graph):
        # Internal switch off — cut output regardless of incoming signal
        if not panel.get('running', True):
            graph.emit(panel, None)
            return

        cap_wh     = panel.get('capacityWh', 84)
        charge_amp = panel.get('chargeAmps', 2)

        if signal is None:
            # No external power — discharge or go off
            panel['inVolts'] = 0.0
            panel['inAmps']  = 0.0
            panel['live']    = False
            if panel.get('chargeWh', 0) > 0 and cap_wh > 0:
                panel['state'] = 'discharging'
                # Emit own EMF as primary source
                graph.emit(panel, {'v': panel.get('volts', 12), 'a': panel.get('amps', 7)})
            else:
                if panel['state'] != 'off':
                    cls.dispatch(panel, 'state:change', {'from': panel['state'], 'to': 'off'})
                panel['state'] = 'off'
                graph.emit(panel, None)
            return

        v_in  = signal.get('v', 0.0)
        a_in  = signal.get('a', 0.0)
        b_v   = panel.get('volts', 12)
        b_a   = panel.get('amps', 7)
        panel['inVolts'] = v_in
        panel['inAmps']  = a_in
        panel['live']    = True

        # Charge contribution
        charge_wh = panel.get('chargeWh', 0.0)

        if not cap_wh or not charge_amp:
            # Passthrough battery
            if panel['state'] != 'pass':
                cls.dispatch(panel, 'state:change', {'from': panel['state'], 'to': 'pass'})
            panel['state']     = 'pass'
            panel['chargeInW'] = 0.0
            graph.emit(panel, {'v': v_in + b_v, 'a': a_in})
            return

        charge_in_w = min(a_in, charge_amp) * b_v
        panel['chargeInW'] = charge_in_w

        if charge_wh >= cap_wh:
            new_state = 'full'
        elif charge_wh <= 0:
            new_state = 'dead'
        elif charge_in_w > 0:
            new_state = 'charging'
        else:
            new_state = 'discharging'

        if panel['state'] != new_state:
            cls.dispatch(panel, 'state:change', {'from': panel['state'], 'to': new_state})
        panel['state'] = new_state

        cls.throttle(panel, 'battery:charge',
                     {'chargePercent': round(panel.get('chargePercent', 0), 1),
                      'chargeWh': round(charge_wh, 3)})

        # Stack battery EMF on top of inbound voltage
        v_out = v_in + b_v
        graph.emit(panel, {'v': round(v_out, 2), 'a': a_in})

    @classmethod
    def tick(cls, panel: Dict, dt: float, graph):
        cap_wh    = panel.get('capacityWh', 0)
        charge_wh = panel.get('chargeWh', 0.0)
        if not cap_wh:
            return

        in_w   = panel.get('chargeInW', 0.0)
        draw_w = panel.get('drawWatts', 0.0)
        b_v    = panel.get('volts', 12)
        total_v = panel.get('inVolts', b_v) + b_v
        out_w   = draw_w * (b_v / total_v) if total_v > 0 else 0.0

        panel['chargeOutW'] = out_w
        delta_wh = (in_w - out_w) * dt / 3600.0
        charge_wh = max(0.0, min(cap_wh, charge_wh + delta_wh))
        panel['chargeWh']      = charge_wh
        panel['chargePercent'] = (charge_wh / cap_wh * 100) if cap_wh else 0

        # Dead / revive transitions
        state = panel['state']
        if state == 'discharging' and charge_wh <= 0:
            panel['state'] = 'dead'
            cls.dispatch(panel, 'battery:dead', {})
            graph.emit(panel, None)
        elif state == 'dead' and in_w > 0 and charge_wh > 0:
            panel['state'] = 'charging'
            cls.dispatch(panel, 'battery:revived', {})

    @classmethod
    def reset(cls, panel: Dict, graph):
        prev               = panel['state']
        cap_wh             = panel.get('capacityWh', 84)
        panel['chargeWh']  = cap_wh * 0.8
        panel['state']     = 'off'
        panel['live']      = False
        panel['running']   = True
        panel['chargeInW'] = 0.0
        panel['chargeOutW'] = 0.0
        cls.dispatch(panel, 'state:change', {'from': prev, 'to': 'off'})
        graph.emit(panel, None)


NodeRegistry.register(SeriesBattery)
